scripts/packages/VSCodeServer/src/eval.jl

```diff
@@ -163,5 +163,23 @@ function backtrace_string(bt)
     end |> joinlines
 end
 
-# HACK: very ugly, this protocol is ...
-vscode_cmd_string(cmd; cmdargs...) = string("vscode-command:", cmd, '?', "argstart", JSON.json(cmdargs), "argend")
+using Printf
+
+UNESCAPED = Set(codeunits("0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_.!~*'()"))
+
+function encode_uri_component(uri)
+    isvalid(uri) || throw(ArgumentError("`encode_uri_component` can only handle valid UTF8 strings."))
+
+    io = IOBuffer()
+    for cp in codeunits(uri)
+        if cp in UNESCAPED
+            print(io, Char(cp))
+        else
+            print(io, '%')
+            @printf(io, "%2X", cp)
+        end
+    end
+    return String(take!(io))
+end
```

Comment:
This is much better, nice.

Reply:
Well, can you tell me any reference of this implementation ? Especially around `@printf(io, "%2X", cp)` ?

Reply:
I have the implementation at https://gist.github.com/pfitzseb/d6a9ebee8404e12ec69132af6cf2dada.

The `@printf(io, "%2X", cp)` thing just prints `cp` as a hex (`x`) string with 2 chars (`2`) and in uppercase (`X`).

Reply:
Got it. Let's add the test examples when we setup the test for this code.